pkg/controllers/infracluster/azure.go

```diff
@@ -0,0 +1,251 @@
+		if err := createNewAzureSecret(ctx, r.Client, capzManagerBootstrapSecret.Data["azure_client_secret"]); err != nil {
+			return nil, fmt.Errorf("failed to create Azure Cluster secret: %w", err)
+		}
+	}
+
+	apiURL, err := url.Parse(r.Infra.Status.APIServerInternalURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse apiUrl: %w", err)
+	}
+
+	port, err := strconv.ParseInt(apiURL.Port(), 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse apiUrl port: %w", err)
+	}
+
+	if r.Infra.Status.PlatformStatus == nil {
+		return nil, fmt.Errorf("infrastructure PlatformStatus should not be nil: %w", err)
+	}
+
+	azureClientID, ok := capzManagerBootstrapSecret.Data["azure_client_id"]
+	if !ok {
+		return nil, fmt.Errorf("failed to get azureClientID: %w", err)
+	}
+
+	azureTenantID, ok := capzManagerBootstrapSecret.Data["azure_tenant_id"]
+	if !ok {
+		return nil, fmt.Errorf("failed to get azureTenantID: %w", err)
+	}
+
+	azureClusterIdentity := &azurev1.AzureClusterIdentity{}
+	// Get the Azure Cluster Identity.
+	if err := r.Get(ctx, client.ObjectKey{Namespace: defaultCAPINamespace, Name: r.Infra.Status.InfrastructureName}, azureClusterIdentity); err != nil && !cerrors.IsAlreadyExists(err) {
+		if !cerrors.IsNotFound(err) {
+			return nil, fmt.Errorf("failed to get Azure Cluster Identity: %w", err)
+		}
+
+		azureClusterIdentity = &azurev1.AzureClusterIdentity{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      r.Infra.Status.InfrastructureName,
+				Namespace: defaultCAPINamespace,
+				Annotations: map[string]string{
+					// The ManagedBy Annotation is set so CAPI infra providers ignore the InfraCluster object,
+					// as that's managed externally, in this case by the cluster-capi-operator's infracluster controller.
+					clusterv1.ManagedByAnnotation: managedByAnnotationValueClusterCAPIOperatorInfraClusterController,
+				},
+			},
+			Spec: azurev1.AzureClusterIdentitySpec{
+				Type:              azurev1.ServicePrincipal,
+				AllowedNamespaces: &azurev1.AllowedNamespaces{NamespaceList: []string{defaultCAPINamespace}},
+				ClientID:          string(azureClientID),
+				TenantID:          string(azureTenantID),
+				ClientSecret:      corev1.SecretReference{Name: clusterSecretName, Namespace: defaultCAPINamespace},
+			},
+		}
+
+		// The Azure Cluster Identtiy does not exist, so it needs to be created.
+		if err := r.Create(ctx, azureClusterIdentity); err != nil && !cerrors.IsAlreadyExists(err) {
+			return nil, fmt.Errorf("failed to create Azure Cluster Identity: %w", err)
+		}
+	}
+
+	target := &azurev1.AzureCluster{ObjectMeta: metav1.ObjectMeta{
+		Name:      r.Infra.Status.InfrastructureName,
+		Namespace: defaultCAPINamespace,
+	}}
+
+	// Checking whether InfraCluster object exists. If it doesn't, create it.
+	if err := r.Get(ctx, client.ObjectKeyFromObject(target), target); err != nil {
+		if !cerrors.IsNotFound(err) {
+			return nil, fmt.Errorf("failed to get InfraCluster: %w", err)
+		}
```

Comment:
If this were factored out into a separate function, I wonder if you could make this less indented by returning early in a nice way?

Reply:
I think this is similar to what I asked above - I'm not sure I see how this could be pulled out on it's own when this is supposed to be the purpose of this function, right? Or am I misunderstanding?

Reply:
You can always factor things into smaller chunks if they help tell the story.

If your top level function reads as four or five "do this bit" and factors into smaller functions to handle the implementation of those, that's a nice way to make the function readable.

We quite often use this pattern where we have a `Reconcile` and a `reconcile` function, where the top level does the boilerplate, and the inner reconcile does more specific business logic.

At the end of the day, the reason we don't like long functions is that they are hard to follow, so this is only about readability and improving the experience for us, 6 months down the line when we can no longer remember what's happening here

Reply:
Yeah, I see what you mean now, I've refactored it further so let me know what you think now whenever you get the chance